primeorder/src/risc0.rs

```diff
@@ -0,0 +1,319 @@
+        result.sqn(2, (&mut scratch_1, &mut scratch_2), scratch);
+        result.mul_unchecked(scratch, &mut scratch_1);
+
+        // scratch_1 = t1111
+        // let t11111111 = t1111.mul(&t1111.sqn(4));
+        scratch_1.sqn(4, (&mut scratch_2, result), scratch);
+        scratch_1.mul_unchecked(scratch, result);
+
+        // result = t11111111
+        // let x16 = t11111111.sqn(8).mul(&t11111111);
+        result.sqn(8, (&mut scratch_1, &mut scratch_2), scratch);
+        result.mul_unchecked(scratch, &mut scratch_1);
+
+        // scratch_1 = x16
+        // let sqrt = x16
+        //     .sqn(16)
+        scratch_1.sqn(16, (&mut scratch_2, result), scratch);
+        //     .mul(&x16)
+        scratch.mul_unchecked(&scratch_1, result);
+        //     .sqn(32)
+        result.sqn(32, (&mut scratch_1, &mut scratch_2), scratch);
+        //     .mul(self)
+        scratch.mul_unchecked(self, &mut scratch_1);
+        //     .sqn(96)
+        scratch_1.sqn(96, (&mut scratch_2, result), scratch);
+        //     .mul(self)
+        scratch.mul_unchecked(self, &mut scratch_2);
+        //     .sqn(94);
+        // Last result is written to the result buffer.
+        scratch_2.sqn(94, (&mut scratch_1, scratch), result);
+    }
+
+    /// Returns self^(2^n) mod p.
+    ///
+    /// This implementation is designed to avoid any memcpy of the buffers for intermediate ops.
+    fn sqn(&self, n: usize, scratch: (&mut Self, &mut Self), result: &mut Self) {
+        let mut x = scratch.0;
+        let mut buffer = scratch.1;
+
+        if n == 1 {
+            // self^(2^1) = self^2
+            self.square(result);
+            return;
+        } else if n == 0 {
+            // self^(1) = self
+            *result = *self;
+            return;
+        }
+
+        // write value to a scratch buffer.
+        self.square(x);
+
+        // Square n - 2 times.
+        let mut i = 2;
+        while i < n {
+            x.square(buffer);
+            i += 1;
+            // Swap scratch buffers, to set x to the squared value.
+            core::mem::swap(&mut x, &mut buffer);
+        }
+
+        // Write final square to result buffer.
+        x.square(result);
+    }
+
+    /// Returns self^2 mod p
+    pub(crate) fn square(&self, result: &mut Self) {
+        self.mul_unchecked(self, result);
+    }
+}
+
+fn bytes_to_u32_words_le(bytes: &[u8]) -> [u32; 8] {
+    let mut words = [0u32; 8];
+
+    // Process 4 bytes at a time to create little-endian u32 words
+    for (i, chunk) in bytes.chunks(4).enumerate() {
+        // Convert each big-endian chunk to a little-endian u32
+        words[7 - i] = u32::from_be_bytes(chunk.try_into().unwrap());
+    }
+
+    words
+}
+
+pub(crate) fn felt_to_u32_words_le<C>(data: &C::FieldElement) -> [u32; 8]
+where
+    C: PrimeCurveParams,
+{
+    bytes_to_u32_words_le(data.to_repr().as_slice())
+}
+
+#[inline]
+fn affine_to_r0_affine<C>(affine: &AffinePoint<C>) -> ec::AffinePoint<8, C>
```

Comment:
Maybe my other comments about testing for the identity actually boil down to this function -- I'm not confident that this function will return appropriately when passed the identity.